cni-plugin/pkg/dataplane/linux/dataplane_linux.go

```diff
@@ -150,8 +150,26 @@ func (d *linuxDataplane) DoNetworking(
 			return fmt.Errorf("failed to set %q up: %w", contVethName, err)
 		}
 
-		// Fetch the MAC from the container Veth. This is needed by Calico.
-		contVethMAC = contVeth.Attrs().HardwareAddr.String()
+		// Check if there is an annotation requesting a specific fixed MAC address for the container Veth, otherwise
+		// use kernel-assigned MAC.
+		requestedContVethMac, found := annotations["cni.projectcalico.org/hwAddr"]
+		if found {
+			tmpContVethMAC, err := net.ParseMAC(requestedContVethMac)
+			if err == nil {
```

Comment:
This isn't necessarily wrong, but it's not a typical error handling pattern in go. Generally better to handle errors immediately rather than try to condense multiple error generating functions into a single error check. I think for most cases it's more idiomatic to do an `err != nil` check than an `err == nil` check as well, although there are obviously exceptions to that. 

e.g.,

```go
mac, err := parseMAC(..)
if err != nil {
  return err
}

err = netlink.LinkSetHardwareAddr(...)
if err != nil {
  return err
}
```

^ this also lets us tailor specific error messages to the user. e.g., if the annotation is bad, we want to say something specifically like "The MAC address provided via cni.projectcalico.org is invalid: %s"

Reply:
That makes sense specifically about having unique messages for each failure case.